### PR TITLE
[agent-e][issue #1632] Retire event receiver projections

### LINKED ISSUE #1632: [runtime/spec] Clean up legacy event receiver projection shorthands
## Context

Child split from #1627 after the lead gate disposition in https://github.com/division-sh/swarm/issues/1627#issuecomment-4871816639.

#1627 remains the parent/design tracker. Full `_event.*` migration is watchlist-only. `event.*` remains the documented platform-owned event context root for handler expressions.

## Problem

`event.entity_id` and `event.flow_instance` are legacy top-level receiver-context projections. They are not source-of-truth route identity, but they look like event identity facts and can be confused with `event.source`, `event.target`, and `event.target_set`.

The source-of-truth route identity model is:

- `event.source` for emitter route identity
- `event.target` for singular/current delivery receiver route identity
- `event.target_set` for persisted fan-out recipient route identities

Current receiver entity context should be read through `_entity.id` / `_entity.flow_instance`, not through top-level `event.entity_id` / `event.flow_instance`.

## Scope

In scope:

- Keep `event.*` as the platform-owned handler event context root.
- Define the exact supported `event.*` field surface in `platform-spec.yaml`.
- Mark top-level `event.entity_id` and `event.flow_instance` unsupported for authoring expression surfaces.
- Migrate existing supported authoring fixtures/tests away from top-level `event.entity_id` / `event.flow_instance`.
- Add fail-closed validation/proof so those legacy projections cannot remain as silent runtime-only fallbacks.
- Preserve source authority: selectors must not gain any `event.*` bypass.

Out of scope:

- No `_event.*` migration.
- No payload/envelope admission migration except proof that this change does not weaken existing guards.
- No persisted event readback, replay, delivery, diagnostics, or dead-letter schema migration unless a same-class failure is discovered during pre-audit.
- No route identity model rewrite beyond retiring the legacy top-level projection pair.

## Candidate supported event fields

Supported handler event context fields should remain approximately:

- `event.id`
- `event.type`
- `event.source`
- `event.target`
- `event.target_set`
- `event.source_event_id`
- `event.emitted_at`
- `event.trigger_event_type`
- `event.current_state`
- `event.run_id`
- `event.scope`

Unsupported top-level legacy projections:

- `event.entity_id`
- `event.flow_instance`

## Replacement guidance

- Current receiver entity id: `_entity.id`
- Current receiver flow instance: `_entity.flow_instance`
- Emitter route entity/flow: `event.source.entity_id` / `event.source.flow_instance`
- Receiver route entity/flow: `event.target.entity_id` / `event.target.flow_instance`
- Fan-out recipient set: `event.target_set`

## Required pre-audit

Before implementation, inventory all current `event.entity_id` / `event.flow_instance` consumers and classify each as:

- moved to `_entity.*` current receiver context
- moved to `event.source.*` route source identity
- moved to `event.target.*` route receiver identity
- different concept, with proof
- split/escalated

The audit must name canonical owner(s), spec updates, validation owners, required negative proof, and whether this child can close in one PR.

## Acceptance expectations

- `platform-spec.yaml` is updated in the implementation PR.
- Top-level `event.entity_id` and `event.flow_instance` fail closed on all supported authoring expression surfaces.
- Existing supported usages are migrated to explicit `_entity.*`, `event.source.*`, or `event.target.*` reads.
- Positive proof covers the supported `event.*` fields that remain.
- Negative proof covers unsupported `event.entity_id` / `event.flow_instance` and selector source-authority bypass attempts.
- Full `go test ./...` passes.

Related: #1627, #1533, PR #1628

### PR DESCRIPTION
**Pull request by yazzaoui**
## Summary

Closes #1632
Part of #1627

Retires the legacy top-level handler event receiver projections `event.entity_id` and `event.flow_instance` from supported authoring expression surfaces while keeping `event.*` as the platform-owned event context root.

## Governing refs / context

Exact spec refs updated in this PR:

- `platform-spec.yaml#contract_formats.event_schema.strict_payload_contract.envelope_split`
- `platform-spec.yaml#handler_specification.expression_context.namespaces.event`
- Adjacent preserved refs: `platform-spec.yaml#contract_formats.event_schema.strict_payload_contract.envelope_identity`, `platform-spec.yaml#handler_specification.expression_context.namespaces._entity`, and selector source-authority sections for `select_entity` / `select_or_create_entity`.

Binding issue/gate context:

- #1632 issue body.
- #1632 Pre-Implementation Coverage Audit.
- #1632 gate comment approved as first slice.
- #1627 remains the broader parent/watchlist for event envelope/root authority ambiguity.

## Semantic migration

New canonical owners:

- `internal/events/context_fields.go`: shared event authoring-field policy for supported and unsupported `event.*` authoring refs.
- `internal/events.Event.ContextMap`: runtime context-map producer, no longer emitting top-level `entity_id` / `flow_instance`.
- `platform-spec.yaml`: authoritative supported `event.*` authoring field surface and unsupported legacy projection guidance.

Old producer / reader / writer path now invalid:

- Top-level authoring refs `event.entity_id` and `event.flow_instance` are unsupported.
- Current receiver context moved to `_entity.id` / `_entity.flow_instance`.
- Route identity reads must use `event.source.*`, `event.target.*`, or `event.target_set` by meaning.

Production paths checked for surviving old behavior:

- Event context production: `Event.ContextMap`.
- Engine direct operand refs and CEL conditions.
- Pipeline workflow CEL/query normalization.
- Workflow value expressions.
- Mailbox/action expression refs and CEL.
- `create_flow_instance config_from` event refs.
- Bootverify condition and emit/action expression validation.
- Selector source-authority guards.

Migration complete for #1632: all in-repo supported authoring usages of `event.entity_id` / `event.flow_instance` are either migrated or retained only as intentional negative proof.

## Proof run

- `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`
- `go test ./internal/runtime/cataloge2e -run 'TestTier1PrimitiveCatalogFixtures_RealRuntime/test-payload-transform-multi-source' -count=1`
- `go test ./internal/runtime/bootverify -run 'TestRun_RejectsSelectEntityByPlatformEntitySourceAuthority|TestRun_RejectsSelectEntityWithSourceEnvelopeAuthority|TestRun_RejectsSelectOrCreateEntityWithSourceEnvelopeAuthority' -count=1`
- `go test ./...`

## Scope notes

This PR does not introduce `_event.*`, does not change persisted event/readback/replay/dead-letter schema, does not alter payload/envelope admission, and does not rewrite route identity semantics. Those remain parent/watchlist territory under #1627 unless a future concrete child is opened.

### CODE REVIEW

**Comment by yazzaoui** [comment]
## Post-Implementation Proof Audit

### Exact Semantic Concepts Changed

- Supported handler `event.*` authoring field surface.
- Unsupported legacy top-level receiver projections: `event.entity_id` and `event.flow_instance`.
- Replacement ownership:
  - current receiver context: `_entity.id` / `_entity.flow_instance`
  - emitter route identity: `event.source.entity_id` / `event.source.flow_instance`
  - singular receiver route identity: `event.target.entity_id` / `event.target.flow_instance`
  - fan-out route identity: `event.target_set`
- Runtime and bootverify now consume the same event authoring-field policy.

### Chosen Working Failure Class Closed

Closed class: legacy top-level handler event receiver projections, specifically `event.entity_id` and `event.flow_instance`, remained live as supported authoring reads instead of failing closed or being replaced by explicit receiver/route context reads.

Immediate parent failure class: event envelope/root authority ambiguity after the `_entity` migration.

Broader parent failure class: platform/business namespace separation and source-authority preservation across expression, routing, payload, and persisted-event surfaces.

### Canonical Owners After Change

- Spec owner: `platform-spec.yaml#contract_formats.event_schema.strict_payload_contract.envelope_split` and `platform-spec.yaml#handler_specification.expression_context.namespaces.event`.
- Runtime policy owner: `internal/events/context_fields.go` (`ValidateEventContextReference`, supported/unsupported event field policy).
- Runtime context producer: `internal/events.Event.ContextMap`, which no longer emits top-level `entity_id` / `flow_instance`.
- Receiver-context owner remains `_entity.*` from `platform-spec.yaml#handler_specification.expression_context.namespaces._entity` and the existing platform entity context implementation.
- Route identity owner remains `event.source`, `event.target`, and `event.target_set`.

### Touched Consumers and Systematic Consumption Audit

| Owner | Consumer | Status |
| --- | --- | --- |
| `platform-spec.yaml` event namespace | Spec lists supported top-level fields, route subfields, unsupported legacy projections, and source-authority constraint | moved to canonical owner in this work |
| `internal/events/context_fields.go` | `workflowexpr.ValidateValueExpression*` and `EvalValueExpression*` | moved to canonical owner in this work |
| `internal/events/context_fields.go` | Pipeline condition/CEL normalization, including query_entities operand normalization | moved to canonical owner in this work |
| `internal/events/context_fields.go` | Engine direct operand resolution and compiled conditions | moved to canonical owner in this work |
| `internal/events/context_fields.go` | Mailbox/action expression refs and CEL values | moved to canonical owner in this work |
| `internal/events/context_fields.go` | `create_flow_instance config_from` event refs | moved to canonical owner in this work |
| `internal/events/context_fields.go` | Bootverify condition and emit/action expression validation through shared expression validators | moved to canonical owner in this work |
| `internal/events.Event.ContextMap` | Runtime handler event map producer | moved to canonical owner in this work by removing legacy top-level projection keys |
| `_entity.*` platform entity context | Current receiver refs in mailbox/API/contract/catalog/pipeline fixtures | already consumes canonical owner; touched consumers migrated in this work |
| `event.source.*` route identity | Source-route proof in executor emit-intent fixture | already consumes canonical owner; touched consumer migrated in this work |
| selector source-authority validators | `select_entity` / `select_or_create_entity` restrictions | already consumes canonical owner; execution-proven unchanged |
| persisted event rows, replay, diagnostics, dead-letter/readback projections | Store/readback projection concepts | different semantic concept; not touched |
| payload/envelope admission | Payload-owned vs platform-owned event fields | different semantic concept; not touched |
| full `_event.*` migration | Broader expression-root cleanup | explicitly split/watchlist under #1627/#1528 |

### Old Paths

Invalid/non-authoritative on supported authoring expression surfaces:

- `event.entity_id`
- `event.flow_instance`

Still authoritative:

- `_entity.id`
- `_entity.flow_instance`
- `event.source.*`
- `event.target.*`
- `event.target_set`
- supported event metadata fields: `event.id`, `event.type`, `event.trigger_event_type`, `event.source_agent`, `event.task_id`, `event.source_event_id`, `event.emitted_at`, `event.current_state`, `event.run_id`, `event.scope`

Still surviving but not authoring-surface authority:

- `Event.EntityID()` / `Event.FlowInstance()` and persisted event identity/projection rows remain internal/projection helpers. They were not changed because persisted readback/replay/dead-letter schema was explicitly outside #1632.

### Broader Failure Class / Symptom Shape / Closure Level

The issue was a child slice of a broader parent, not the whole parent. It was symptom-shaped in its original local manifestations, but the approved class boundary was coherent: remove the two top-level receiver projection shorthands from all supported authoring expression surfaces.

Achieved closure level: chosen child class eliminated. Top-level `event.entity_id` / `event.flow_instance` are no longer produced by the handler event context map and fail closed through runtime and bootverify policy consumers.

Parent remains open: yes. #1627 remains the parent/watchlist/design tracker for event envelope/root authority ambiguity and any future `_event`, payload/envelope admission, persisted readback/replay/dead-letter, or route-identity-model work.

Remaining child-slice tail estimate for parent: unchanged from pre-audit. No new concrete same-class child was discovered during implementation. Broader work remains watchlist-only unless a concrete failure appears. Implementation did identify the expected architecture smell, but did not change the parent estimate.

### Sibling Contexts Checked

- `_event` full migration: explicitly out of scope and not introduced.
- Payload/envelope admission: no code changes; existing envelope-owned payload guards still covered by bootverify tests.
- Persisted event/readback/replay/dead-letter/diagnostics: no files touched; different concept.
- Route identity storage/model: no rewrite; `event.source`, `event.target`, `event.target_set` remain canonical route identity.
- Selector source authority: explicitly re-run and preserved.
- Mailbox/action refs: checked and migrated to `_entity.*` where receiver context was intended.
- Catalog fixture: checked and migrated to `_entity.id`.

### Generic Failing Proof

Created generic policy/runtime proof rather than fixture-only proof:

- `internal/events` tests prove `ContextMap` omits legacy projection keys and the policy rejects the retired names.
- `internal/runtime/workflowexpr` tests prove value-expression validation and eval reject legacy refs even if a caller manually supplies the old map key.
- `internal/runtime/pipeline` tests prove condition validation rejects legacy refs, including inside `query_entities(...)` operands.
- `internal/runtime/engine` tests prove direct operand and compiled condition paths reject legacy refs.
- `internal/runtime/bootverify` tests prove condition and emit-field authoring refs fail closed.

### Watchlist Decision

#1627 remains the parent/watchlist for broader event envelope/root authority ambiguity. #1528 remains the broader authoring-surface simplification parent. No new child issue is required from this implementation because no additional live same-concept authoring interpreter was found beyond the approved boundary.

### Architecture Issue / Tracking / Effort ROI

Architecture smell: event authoring-field support was implicit across `Event.ContextMap`, dynamic CEL maps, direct ref lookups, and bootverify validation. This PR adds a minimal shared policy owner consumed by the touched runtime/bootverify seams. Long-run direction is still a broader expression-root registry if more roots show the same drift pattern.

Tracking decision: current minimal policy is fixed in #1632; broader registry/full `_event` migration stays watchlist-only under #1627/#1528. Estimated broader effort: medium to high; ROI is conditional on another concrete root-surface drift manifestation. Current #1632 ROI was high because it removed copied local interpretations for the approved child class.

### Manifestation Coverage Table

| Known manifestation | Status | Exact proof |
| --- | --- | --- |
| `Event.ContextMap` exposed top-level `entity_id` / `flow_instance` | reproduced and fixed | `TestEventContextMapOmitsLegacyReceiverProjectionFields`; focused `internal/events` package run; `go test ./...` |
| Engine direct operand and compiled condition paths accepted legacy event refs | reproduced and fixed | `TestExecutionScopeResolveOperand_RejectsLegacyEventReceiverProjection`; `TestCompiledExecutionCondition_RejectsLegacyEventReceiverProjection`; focused engine package run; `go test ./...` |
| Workflow value expressions accepted/evaluated legacy event refs | reproduced and fixed | `TestValidateValueExpression_RejectsLegacyEventReceiverProjections`; `TestEvalValueExpression_RejectsLegacyEventReceiverProjectionEvenWhenMapContainsValue`; focused workflowexpr package run; `go test ./...` |
| Pipeline condition/CEL/query normalization accepted legacy event refs | reproduced and fixed | `TestValidateConditionCEL_RejectsLegacyEventReceiverProjections`; focused pipeline package run; `go test ./...` |
| Bootverify allowed conditions and emit/action refs with legacy event projections | reproduced and fixed | `TestRun_RejectsLegacyEventReceiverProjectionReferences`; focused bootverify package run; `go test ./...` |
| Mailbox/API/action supported fixtures authored receiver refs through `event.entity_id` / `event.flow_instance` | execution-proven through the same corrected path | migrated to `_entity.*`; focused pipeline and apiv1 package run; `go test ./...` |
| Catalog fixture `test-payload-transform-multi-source` authored receiver id through `event.entity_id` | execution-proven through the same corrected path | migrated to `_entity.id`; cataloge2e tier1 subtest for `test-payload-transform-multi-source`; `go test ./...` |
| Selector source-authority bypass risk from event/context roots | execution-proven through the same corrected path | `TestRun_RejectsSelectEntityByPlatformEntitySourceAuthority`; `TestRun_RejectsSelectEntityWithSourceEnvelopeAuthority`; `TestRun_RejectsSelectOrCreateEntityWithSourceEnvelopeAuthority`; `go test ./...` |
| Persisted readback/replay/dead-letter/payload admission/full `_event` migration | split / escalated as separate class | explicitly not touched; parent/watchlist remains #1627/#1528 |

### Required Supported-Surface Proof Actually Run

- `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`
- `go test ./internal/runtime/cataloge2e -run 'TestTier1PrimitiveCatalogFixtures_RealRuntime/test-payload-transform-multi-source' -count=1`
- `go test ./internal/runtime/bootverify -run 'TestRun_RejectsSelectEntityByPlatformEntitySourceAuthority|TestRun_RejectsSelectEntityWithSourceEnvelopeAuthority|TestRun_RejectsSelectOrCreateEntityWithSourceEnvelopeAuthority' -count=1`
- `go test ./...`

### Optional General Feedback

None closure-bearing beyond the architecture smell above.

File: internal/runtime/workflowexpr/data_expression.go [hunk 1]
```diff
@@ -22,6 +23,7 @@ var (
 
 	workflowExpressionEntityReferencePattern         = regexp.MustCompile(`(^|[^a-zA-Z0-9_])entity\.([a-zA-Z_][a-zA-Z0-9_.]*)`)
 	workflowExpressionPlatformEntityReferencePattern = regexp.MustCompile(`(^|[^a-zA-Z0-9_])_entity\.([a-zA-Z_][a-zA-Z0-9_.]*)`)
+	workflowExpressionEventReferencePattern          = regexp.MustCompile(`(^|[^a-zA-Z0-9_])event\.([a-zA-Z_][a-zA-Z0-9_.]*)`)
```
[the remark this answers is not in the document]
Comment (yazzaoui) [reply internal/runtime/workflowexpr/data_expression.go]: Addressed in 97083908. `EventReferences` now only treats `event.` as a platform event root when it appears at a real root boundary; nested author data such as `payload.event.entity_id`, `_entity.event.flow_instance`, and spaced member access are ignored by the event-context validator. Added regression coverage for both root matches and nested author-field non-matches.

Proof run after the fix:
- `go test ./internal/runtime/workflowexpr -count=1`
- `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`
- `go test ./...`

File: internal/runtime/workflowexpr/data_expression.go [hunk 2]
```diff
@@ -332,6 +360,72 @@ func PlatformEntityReferences(expression string) []string {
...
+		refStart := start + len(eventPrefix)
+		searchStart = refStart
+		if !isRootReferenceStart(expression, start) {
+			continue
+		}
+		if refStart >= len(expression) || !isIdentifierStart(expression[refStart]) {
+			continue
+		}
+		refEnd := refStart + 1
+		for refEnd < len(expression) && (isIdentifierPart(expression[refEnd]) || expression[refEnd] == '.') {
```
Comment (yazzaoui) [inline internal/runtime/workflowexpr/data_expression.go]: Blocking: this scanner only collects dotted `event.<identifier>` paths, so CEL map-index forms like `event["entity_id"]` and `event["flow_instance"]` do not reach `events.ValidateEventContextReference`. Those are still top-level reads of the retired receiver projection keys, and `ValidateValueExpression` / bootverify emit-field validation will compile them instead of failing closed through the shared policy. Please either reject event-root map indexing for unsupported keys or route those accesses through the same policy, and add negative proof for the bracket forms on the relevant CEL authoring surfaces.

**Review by yazzaoui — COMMENTED** [review]
CHANGES NEEDED.

The pre-audit/gate framing is valid and the implementation covers the normal dotted `event.*` paths, but the closure claim is not yet defensible because CEL map-index access still bypasses the event-field policy. That leaves the same retired top-level receiver projections authorable on CEL value/emit surfaces without bootverify/runtime policy rejection.

The prior inline thread about nested author `payload.event.*` was resolved and is obsolete on current head. This is a separate bypass form under the same closure class.

**Comment by yazzaoui** [comment]
## Review checklist

Status: `CHANGES NEEDED`.

- Process gates: PR description, issue #1632, issue thread, #1627 parent tracker, #1528 broader tracker, pre-audit, Gate E approval, PR proof audit, PR comments, and inline threads read.
- Unresolved comments: prior inline thread is resolved/outdated and verified obsolete on current head.
- Category: semantic-drift / runtime-spec / route-authority cleanup is correct; independent pre-audit gate exists and approved #1632 as first slice.
- Spec compliance: incomplete. Dotted `event.entity_id` / `event.flow_instance` are retired, but CEL bracket access to the same top-level keys is not policy-validated.
- Failure-class coverage: incomplete until `event["entity_id"]` / `event["flow_instance"]` and equivalent bracket forms fail closed or are routed through the shared policy.
- Runtime/spec authority: `platform-spec.yaml` was updated in this PR; current issue is implementation proof gap, not missing spec artifact.
- Residual risk: high for the chosen closure class because one supported CEL authoring form can bypass the new event-field scanner.
- Risk level: `medium-high` until fixed; expected to become low after bracket-form negative proof lands.
- Closure level achieved now: dotted touched seams canonicalized; chosen child class not fully eliminated.
- Watchlist decision: no new watchlist node; this remains same-class #1632 work under #1627/#1528.
- Merge recommendation: do not merge until the bracket-access bypass is fixed and proven with bootverify/runtime CEL tests.

File: internal/runtime/workflowexpr/data_expression.go (hunk 2 above)
Comment (yazzaoui) [reply internal/runtime/workflowexpr/data_expression.go]: Addressed in ed9a8f1b. The shared event reference scanner now routes root `event` bracket access through the same `events.ValidateEventContextReference` policy as dotted access, including mixed forms such as `event["source"].entity_id` and `event.target["flow_instance"]`.

I also rejected dynamic root `event[...]` indexing because it cannot be statically checked against the exact supported event-field surface; allowing it would preserve the same bypass class through a computed key.

Architecture check: this is same-class #1632 leakage, not a new parent obligation. The local cause was the existing lexical expression scanner only understanding dotted root refs. The broader architecture smell remains the one already tracked under #1627/#1528: root-field validation would be stronger as an AST/root-access registry. For this PR boundary, the canonical policy is still `internal/events/context_fields.go`, and all touched CEL validation consumers now call it through the shared scanner.

Proof run after the fix:
- `go test ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/bootverify -count=1`
- `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`
- `go test ./...`

**Comment by yazzaoui** [comment]
## Post-Implementation Proof Audit Addendum

Review found one additional manifestation in the same #1632 chosen class: CEL bracket access to retired top-level receiver projections (`event["entity_id"]`, `event["flow_instance"]`) bypassed the dotted `event.*` scanner.

### Classification

- Semantic concept: supported handler `event.*` authoring field surface, including equivalent CEL map-index syntax at the root `event` context.
- Chosen working failure class: unchanged. This is still the legacy top-level handler event receiver projection class, not a new parent class.
- Parent class: unchanged. #1627/#1528 continue tracking broader event/root authority and expression-root registry concerns.
- Architecture assessment: the bypass is a symptom of the known local architecture smell that root-field validation is lexical rather than AST/registry-owned. It does not require widening this PR because the corrected shared scanner now routes dotted, bracket, and mixed root `event` accesses through the same canonical event policy. A full AST/root-access registry remains broader watchlist work under #1627/#1528.

### Manifestation Coverage Addendum

| Known manifestation | Status | Exact proof |
| --- | --- | --- |
| CEL bracket reads `event["entity_id"]` / `event["flow_instance"]` bypassed the event-field policy | reproduced and fixed | `TestValidateValueExpression_RejectsLegacyEventReceiverProjections`; `TestValidateValueExpression_RejectsUnsupportedEventBracketRefs`; `TestValidateConditionCEL_RejectsLegacyEventReceiverProjections`; `TestCompiledExecutionCondition_RejectsLegacyEventReceiverProjection`; `TestRun_RejectsLegacyEventReceiverProjectionReferences`; `go test ./...` |
| Mixed root route forms such as `event["source"].entity_id` and `event.target["flow_instance"]` need to consume the same policy | execution-proven through the same corrected path | `TestEventReferences_OnlyMatchesRootEventContext`; `TestValidateValueExpression_AllowsSupportedEventBracketRefs`; focused supported-surface package run; `go test ./...` |
| Dynamic root event indexing `event[payload.key]` could preserve the same bypass with a computed key | reproduced and fixed | dynamic root event indexing now fails closed; `TestValidateValueExpression_RejectsLegacyEventReceiverProjections`; `TestValidateValueExpression_RejectsUnsupportedEventBracketRefs`; bootverify condition negative case; `go test ./...` |

### Proof Run

- `go test ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/bootverify -count=1`
- `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`
- `go test ./...`

**Review by yazzaoui — COMMENTED** [review]
CHANGES NEEDED.

The prior bracket-access bypass for retired `event.entity_id` / `event.flow_instance` is materially addressed in `ed9a8f1b`: dotted, bracket, mixed, and dynamic root `event[...]` forms now route through the shared event-field policy in the workflow expression scanner. However, the follow-up introduced a verify-vs-runtime parity gap for the supported static bracket route forms the PR now claims to support.

Blocking finding: `internal/runtime/pipeline/workflow_condition_validation.go` recognizes `event.` as a condition prefix, but not `event[`. I verified with a temporary local probe that `ValidateConditionCEL("event[\"source\"].entity_id != \"\"", WorkflowConditionContextRule)` succeeds while `WorkflowConditionMissingRecognizedPrefix` returns true. Since bootverify calls that prefix gate in `internal/runtime/bootverify/workflow_expression_checks.go`, a static supported event route read can pass runtime/pipeline CEL validation while bootverify rejects it as `condition_expression_validation: missing required prefix` before semantic validation.

Please make the recognized-prefix gate understand static bracket event roots, or otherwise explicitly stop claiming bracket route forms are supported on condition surfaces and align runtime/verify behavior accordingly. Add bootverify/regression proof for the chosen behavior.

**Comment by yazzaoui** [comment]
## Review checklist

Status: `CHANGES NEEDED`.

- Process gates: PR description, issue #1632, issue thread, pre-audit, Gate E approval, #1627 parent tracker, #1528 broader tracker, PR proof audit, addendum, PR comments, reviews, and inline threads read.
- Unresolved comments: prior blocker thread is resolved/outdated and was rechecked on current head `ed9a8f1b`.
- Category: semantic-drift / runtime-spec / route-authority cleanup remains correct; independent pre-audit gate exists and approved #1632 as first slice.
- Previous blocker: fixed for the retired projection class. Dotted `event.entity_id` / `event.flow_instance`, static bracket forms, mixed route forms, and dynamic `event[...]` are now routed through the shared scanner/policy path.
- New blocker: verify-vs-runtime parity gap. `ValidateConditionCEL` accepts static bracket route reads such as `event["source"].entity_id`, but bootverify's missing-prefix gate only recognizes `event.` and rejects the same expression as missing a required prefix.
- Spec compliance: incomplete until the PR either supports static bracket route event refs consistently on condition surfaces or explicitly rejects that syntax consistently across runtime and verify.
- Failure-class coverage: the chosen child class is mostly closed, but the PR addendum's supported-bracket claim is not fully proof-backed across bootverify/runtime condition gates.
- Runtime/spec authority: `platform-spec.yaml` was updated in this PR; current issue is implementation parity/proof gap, not a missing spec artifact.
- Residual risk: medium. The legacy projection bypass is closed, but a supported static CEL form can still produce false verify failures.
- Risk level: `medium` until the prefix gate and bootverify proof are repaired.
- Closure level achieved now: legacy projection scanner bypass fixed; chosen child class nearly eliminated, but PR is not merge-ready because the corrected surface is inconsistent across verify/runtime.
- Watchlist decision: no new watchlist node; broader event/root registry work remains captured by #1627/#1528.
- Merge recommendation: do not merge until static bracket event route references are either verify-accepted with proof or rejected consistently at all relevant surfaces.

**Comment by yazzaoui** [comment]
Addressed the verify/runtime parity blocker in `38ec2f5e`.

What changed:
- `WorkflowConditionMissingRecognizedPrefix` now recognizes condition roots by root/member syntax rather than literal `root.` substrings.
- Static bracket roots such as `event["source"].entity_id`, `payload["score"]`, and `entity["revision_count"]` are accepted by the prefix gate consistently with `ValidateConditionCEL`.
- Added direct prefix-gate regression tests and changed the bootverify event-context allow test to use `event["source"].entity_id`, proving bootverify no longer rejects that supported static bracket route read as `missing required prefix`.

Architecture check:
- Same #1632 surface/parity gap, not a new parent class.
- The deeper smell remains the known lexical validation layering around expression roots; broader AST/root-access registry work remains tracked under #1627/#1528.
- This PR boundary is now aligned for the bracket event route form it claims to support.

Proof run:
- `go test ./internal/runtime/pipeline ./internal/runtime/bootverify -count=1`
- `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`
- `go test ./...`

**Review by yazzaoui — COMMENTED** [review]
No blocking findings on rereview of current head `38ec2f5e`.

I rechecked both prior blockers. The retired projection bypass is closed: dotted, static bracket, mixed route, and dynamic root `event[...]` forms now reach the shared event-field policy, and `event.entity_id` / `event.flow_instance` fail closed across the relevant runtime/bootverify paths. The verify-vs-runtime parity blocker is also addressed: `WorkflowConditionMissingRecognizedPrefix` now recognizes root/member syntax rather than only literal `root.` prefixes, and bootverify proof now covers `event["source"].entity_id` on a condition surface.

Architecture disposition: the PR does still show the broader lexical/root-recognition smell, but I do not see it hiding another same-class interpreter for #1632 after `38ec2f5e`. For this PR boundary, the remaining architecture debt is honestly tracked by #1627/#1528 rather than needing a new child or widening this PR. The closure-bearing duplicated event-root interpretation that caused the previous blocker is aligned for the supported bracket route form this PR claims to support.

Merge recommendation: merge once normal branch policy is satisfied.

**Comment by yazzaoui** [comment]
## Review checklist

Status: `MERGE RECOMMENDED`.

- Process gates: PR description, issue #1632, issue thread, pre-audit, Gate E approval, #1627 parent tracker, #1528 broader tracker, PR proof audit, addendum, latest remediation comment, PR comments, reviews, and inline threads read.
- Unresolved comments: prior inline threads are resolved/outdated and independently rechecked on current head `38ec2f5e`.
- Category: semantic-drift / runtime-spec / route-authority cleanup remains correct; independent pre-audit gate exists and approved #1632 as first slice.
- Spec compliance: complete for this child. `platform-spec.yaml` defines supported `event.*` authoring fields and marks top-level `event.entity_id` / `event.flow_instance` unsupported.
- Prior blocker status: fixed. Legacy receiver projections fail closed for dotted/static bracket forms, and dynamic root `event[...]` is rejected where it cannot be statically policy-checked.
- Verify/runtime parity: fixed. Static bracket event route refs such as `event["source"].entity_id` are no longer accepted by runtime validation while rejected by bootverify's missing-prefix gate.
- Failure-class coverage: chosen child class eliminated; no missed same-concept authoring interpreter found in repo probes.
- Concept-wide drift reduction: material. `Event.ContextMap`, workflowexpr, pipeline, engine, mailbox/action refs, create_flow config refs, and bootverify now consume the shared event-field policy or migrate to `_entity.*` / route owners.
- Closure level: failure class eliminated for #1632; parent #1627 remains open for broader event/root authority and possible `_event`/AST-root-registry decisions.
- Architecture disposition: existing tracker. The lexical/root-recognition smell remains broader architecture debt under #1627/#1528; no new closure-bearing architecture issue surfaced in this rereview.
- Residual risk: low, limited to broader tracked root-registry design work outside this child.
- Watchlist decision: no watchlist change; #1627/#1528 already capture the remaining architecture direction.
- Follow-up tracking: no new issue created; no `docs/POTENTIAL_ISSUES.md` entry needed.
- Tests run locally: `go test ./internal/runtime/pipeline ./internal/runtime/bootverify -count=1`; temporary prefix edge-case probe; `go test ./internal/events ./internal/runtime/workflowexpr ./internal/runtime/pipeline ./internal/runtime/engine ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/apiv1 -count=1`; catalog fixture subtest; selector/source-authority bootverify subtests; `go test ./...`.
- Merge recommendation: merge once normal branch policy is satisfied.
